corehq/apps/reports_core/filters.py

```diff
@@ -55,9 +58,23 @@ def get_value(self, request_params, user=None):
             kwargs.update(
                 {param.name: request_params[param.name] for param in self.params if param.name in request_params}
             )
-            return self.value(**kwargs)
+            choice_list = self.value(**kwargs)
         else:
-            return self.default_value(**kwargs)
+            choice_list = self.default_value(**kwargs)
+        if (user
+                and domain
+                and self.is_location_filter()
+                and toggles.LOCATION_RESTRICTED_SCHEDULED_REPORTS.enabled(domain)):
+            user_location_ids = user.get_location_ids(domain)
+            if user_location_ids and not user.has_permission(domain, 'access_all_locations'):
+                filtered_choice_list = [choice for choice in choice_list if choice.value in user_location_ids]
```

Comment:
When I suggested moving the filtering into `get_value`, I meant into the `get_value` method specific to the location-safe filter (or perhaps the `value` method?).  That is, inside `LocationDrilldownFilter`, which [already does apply location safety](https://github.com/dimagi/commcare-hq/blob/f383e27fc2d991c39f66e92d6ca372f4dd475a82/corehq/apps/reports_core/filters.py#L479-L485) to the default values, though it looks like it only accounts for the user's primary location ID, instead of all of them.  It doesn't check for access in `value`, though.  Nice thing about that is that class already works with a queryset, and it'll handle descendants easily enough.  As written, this only allows users to select their assigned locations, but they should also be able to select descendants of their assigned locations.

And for `DynamicChoiceListFilter`, it's the choice provider that provides the safety, so the logic can go in there.  I believe Charl is working on a fix for that already in https://github.com/dimagi/commcare-hq/pull/33421

Reply:
Added check in the `value` method if `selected_location_id` belongs to user's any of the locations.